File: orchestrator/app/tools/r_scripts/tximport_summarize.R
#!/usr/bin/env Rscript
# Real tximport pipeline (Bioconductor), verbatim the package's own
# documented workflow (tximport() with a transcript-to-gene map).
# Transcript-to-gene mapping is fetched live from Ensembl via biomaRt
# for exactly the transcript IDs present in the uploaded quant files --
# no baked-in tx2gene reference snapshot. Same Rscript-subprocess
# pattern as seurat_analyze.R -- see app/tools/tximport_summarize.py.
#
# Args: <quant_root_dir> <output_csv>
# quant_root_dir must contain one subdirectory per sample, each with
# either a Salmon quant.sf or a Kallisto abundance.h5/abundance.tsv.
suppressMessages({
  library(tximport)
  library(biomaRt)
})

args <- commandArgs(trailingOnly = TRUE)
quant_root <- args[1]
output_file <- args[2]

sample_dirs <- list.dirs(quant_root, recursive = FALSE)
if (length(sample_dirs) == 0) sample_dirs <- quant_root  # single sample uploaded directly

salmon_files <- file.path(sample_dirs, "quant.sf")
kallisto_h5 <- file.path(sample_dirs, "abundance.h5")

if (all(file.exists(salmon_files))) {
  quant_files <- salmon_files
  quant_type <- "salmon"
} else if (all(file.exists(kallisto_h5))) {
  quant_files <- kallisto_h5
  quant_type <- "kallisto"
} else {
  stop("Could not find a consistent set of quant.sf (Salmon) or abundance.h5 (Kallisto) files under each sample subdirectory.")
}
names(quant_files) <- basename(sample_dirs)

if (quant_type == "salmon") {
  tx_ids <- read.delim(quant_files[1])$Name
} else {
  suppressMessages(library(rhdf5))
  tx_ids <- as.character(h5read(quant_files[1], "aux/ids"))
}
tx_ids_clean <- sub("\\..*$", "", tx_ids)  # strip Ensembl version suffix

mart <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
tx2gene_map <- getBM(
  attributes = c("ensembl_transcript_id", "ensembl_gene_id", "hgnc_symbol"),
  filters = "ensembl_transcript_id",
  values = tx_ids_clean,
  mart = mart
)
tx2gene <- data.frame(
  TXNAME = tx_ids,
  GENEID = tx2gene_map$hgnc_symbol[match(tx_ids_clean, tx2gene_map$ensembl_transcript_id)]
)
tx2gene$GENEID[is.na(tx2gene$GENEID) | tx2gene$GENEID == ""] <- tx2gene$TXNAME[is.na(tx2gene$GENEID) | tx2gene$GENEID == ""]

txi <- tximport(quant_files, type = quant_type, tx2gene = tx2gene, ignoreTxVersion = TRUE)

out <- data.frame(gene = rownames(txi$counts), txi$counts)
write.csv(out, output_file, row.names = FALSE)

File: orchestrator/app/tools/tximport_summarize.py
"""A real tximport MCP tool (docs/17-remaining-tools-wiring-plan.md
Phase 3, R/Bioconductor bridge) -- subprocess-wrapped `Rscript`
running tximport's own canonical workflow, same pattern as
`seurat_analyze`. Transcript-to-gene mapping is fetched live from
Ensembl (via biomaRt, inside the R script) for exactly the transcript
IDs in the uploaded quant files -- no baked-in tx2gene snapshot.
Expects a real uploaded archive containing one Salmon/Kallisto quant
subdirectory per sample (classified as `quant_dir_bundle` by
app/file_uploads.py).
"""
import asyncio
import csv
import io
import subprocess
from pathlib import Path
from typing import Any

from claude_agent_sdk import create_sdk_mcp_server, tool

from app.experiment_context import uploads_dir
from app.file_uploads import classify_upload, extract_bundle

R_SCRIPT = str(Path(__file__).parent / "r_scripts" / "tximport_summarize.R")
MAX_ROWS_RETURNED = 30


def _run_tximport(quant_root: Path, out_path: Path) -> tuple[int, str, str]:
    proc = subprocess.run(
        ["Rscript", R_SCRIPT, str(quant_root), str(out_path)],
        capture_output=True, text=True, timeout=900,
    )
    return proc.returncode, proc.stdout, proc.stderr


@tool(
    "tximport_summarize_quants",
    "Given the filename of a real uploaded archive containing one "
    "Salmon or Kallisto quantification subdirectory per sample (see "
    "list_uploaded_files -- format `quant_dir_bundle`), run tximport "
    "to summarize transcript-level quantifications to gene-level "
    "counts, using a transcript-to-gene mapping looked up live from "
    "Ensembl. Never state a gene count this tool didn't actually "
    "compute.",
    {"filename": str},
)
async def tximport_summarize_quants(args: dict[str, Any]) -> dict[str, Any]:
    filename = (args.get("filename") or "").strip()
    updir = uploads_dir()
    if not filename or updir is None:
        return {"content": [{"type": "text", "text": "filename must be non-empty, and this experiment must have an uploaded file."}]}
    file_path = updir / filename
    if not file_path.is_file():
        return {"content": [{"type": "text", "text": f"No uploaded file named {filename!r} in this experiment -- call list_uploaded_files first."}]}

    file_format = classify_upload(file_path)
    if file_format != "quant_dir_bundle":
        return {"content": [{"type": "text", "text": f"{filename!r} was detected as format `{file_format}`, not a Salmon/Kallisto quant directory bundle."}]}
    try:
        quant_root = extract_bundle(file_path)
    except (ValueError, Exception) as exc:  # noqa: BLE001
        return {"content": [{"type": "text", "text": f"Could not extract {filename!r}: {exc}"}]}

    out_path = updir / f"{filename}.tximport_genes.csv"
    code, out, err = await asyncio.to_thread(_run_tximport, quant_root, out_path)
    result_text = out_path.read_text() if out_path.exists() else ""

    if code != 0 or not result_text.strip():
        return {"content": [{"type": "text", "text": f"tximport summarization failed: {err.strip()[-1500:] or out.strip()[-1500:] or 'unknown error'}"}]}

    rows = list(csv.DictReader(io.StringIO(result_text)))
    if not rows:
        return {"content": [{"type": "text", "text": f"tximport found no gene-level rows in {filename}."}]}

    sample_cols = [c for c in rows[0].keys() if c != "gene"]
    lines = [f"tximport gene-level summarization for {filename} [tximport:gene] -- {len(rows)} gene(s) across sample(s): {', '.join(sample_cols)}"]
    for row in rows[:MAX_ROWS_RETURNED]:
        counts = ", ".join(f"{c}={row.get(c, '?')}" for c in sample_cols)
        lines.append(f"- {row.get('gene', '?')}: {counts}")
    if len(rows) > MAX_ROWS_RETURNED:
        lines.append(f"... and {len(rows) - MAX_ROWS_RETURNED} more gene(s) not shown.")

    return {"content": [{"type": "text", "text": "\n".join(lines)}]}


def build_tximport_summarize_mcp_server():
    return create_sdk_mcp_server(name="tximport_summarize", tools=[tximport_summarize_quants])
